Autogen shorthand parser dispatch utilizing information shared with extractor

weinig opened the pull request:
#### 3615b79bbf03b4b1ffc8644d02e5f74bca37c0e6
<pre>
Autogen shorthand parser dispatch utilizing information shared with extractor
<a href="https://bugs.webkit.org/show_bug.cgi?id=293161">https://bugs.webkit.org/show_bug.cgi?id=293161</a>

Reviewed by Darin Adler.

Autogenerates the dispatch switch statement for style property shorthand
parsers, moving the structure of CSS property parser to more closely align
with the the other autogenerated types.

To allow the shorthand parsers to be extracted from CSSPropertyParser,
the result Vector and the member functions to mutate it, were bundled
up into a new CSS::PropertyParserResult type. Then, the individual consumer
functions were moved to CSSPropertyParserCustom.h and made static functions
that take the new result type by reference.

With the shorthand consumers out of CSSPropertyParser, it became clear that
actually having the class have member data was unnecessary, and instead,
the remaining member functions, which were all private, could instead be
moved into CSSPropertyParser.cpp and passed the data they needed. This allowed
simplifying the includes significantly.

Additionally, the functions to compute property initial values were out
of place and so were moved to their own file, CSSPropertyInitialValues.h/cpp.
It should eventually be autogenerated as well.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::isCustomIdentValue):
    - Moved from both ShorthandSerializer.cpp and CSSPropertyParser.cpp.

* Source/WebCore/css/CSSProperties.json:

* Source/WebCore/css/CSSPropertyInitialValues.cpp: Added.
* Source/WebCore/css/CSSPropertyInitialValues.h: Added.
    - Moved from CSSPropertyParser.h/cpp

* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StylePropertiesInlines.h:
    - Add #include for CSSPropertyInitialValues.h

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
    - Removes all member data and functions, instead using static functions
      in CSSPropertyParser.cpp.

* Source/WebCore/css/parser/CSSPropertyParserCustom.h: Added.
    - Shorthand consumers moved here from CSSPropertyParser.cpp so they
      be called from generated dispatch function.

* Source/WebCore/css/parser/CSSPropertyParserResult.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserResult.h: Added.
    - Encapsulates the parser result vector and mutation functions.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Support generating the shorthand parser dispatch. Utilize the new
      CSSPropertyParserCustom.h to also include all the parser helper
      headers, allowing us to avoid needing to update the script when
      it changes.

* Source/WebCore/style/StyleScopeRuleSets.cpp:
    - Add needed #include.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.h:
    - Update tests and results for new shorthand parser support properties and generation.

Canonical link: <a href="https://commits.webkit.org/295311@main">https://commits.webkit.org/295311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e39e6d7c4668d1ed974bc89d8396645d10eadc3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104728 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109942 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/55402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32986 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/79520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/55402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107734 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/19307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94522 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/59827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/104206 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/19075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12599 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/54779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112337 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31893 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/23458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/88602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32257 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88226 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22478 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33117 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/10893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31818 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37171 "Build is in progress. Recent messages:") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/31610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/34951 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33169 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSPropertyInitialValues.cpp:
```diff
@@ -0,0 +1,622 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Copyright (C) 2016-2022 Apple Inc. All rights reserved.
```
I wrote all the code that was moved into this file more recently, so I think this should not have Chromium copyright and should have a newer date. Also should have the Apple Inc rather than Google Inc version of the header.

darinadler commented on Source/WebCore/css/CSSPropertyInitialValues.h:
```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (C) 2016-2025 Apple Inc. All rights reserved.
```
I think this code is all from 2023 and newer:
```suggestion
 * Copyright (C) 2023-2025 Apple Inc. All rights reserved.
```

darinadler commented on Source/WebCore/css/CSSPropertyInitialValues.cpp:
```diff
@@ -0,0 +1,617 @@
+/*
+ * Copyright (C) 2016-2025 Apple Inc. All rights reserved.
```
I think this code is all from 2023 and newer:
```suggestion
 * Copyright (C) 2023-2025 Apple Inc. All rights reserved.
```

darinadler commented on Source/WebCore/css/CSSPropertyInitialValues.cpp:
```diff
@@ -0,0 +1,617 @@
+
+static bool isNumber(const RectBase& quad, double number, CSSUnitType type)
+{
+    return isNumber(quad.protectedTop(), number, type)
+        && isNumber(quad.protectedRight(), number, type)
+        && isNumber(quad.protectedBottom(), number, type)
+        && isNumber(quad.protectedLeft(), number, type);
```
We use `protectedTop` here but just `top` in `isValueID`. Not that we need to fix this, just observing this inconsistency.

weinig committed:
Autogen shorthand parser dispatch utilizing information shared with extractor
https://bugs.webkit.org/show_bug.cgi?id=293161

Reviewed by Darin Adler.

Autogenerates the dispatch switch statement for style property shorthand
parsers, moving the structure of CSS property parser to more closely align
with the the other autogenerated types.

To allow the shorthand parsers to be extracted from CSSPropertyParser,
the result Vector and the member functions to mutate it, were bundled
up into a new CSS::PropertyParserResult type. Then, the individual consumer
functions were moved to CSSPropertyParserCustom.h and made static functions
that take the new result type by reference.

With the shorthand consumers out of CSSPropertyParser, it became clear that
actually having the class have member data was unnecessary, and instead,
the remaining member functions, which were all private, could instead be
moved into CSSPropertyParser.cpp and passed the data they needed. This allowed
simplifying the includes significantly.

Additionally, the functions to compute property initial values were out
of place and so were moved to their own file, CSSPropertyInitialValues.h/cpp.
It should eventually be autogenerated as well.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
(WebCore::isCustomIdentValue):
    - Moved from both ShorthandSerializer.cpp and CSSPropertyParser.cpp.

* Source/WebCore/css/CSSProperties.json:

* Source/WebCore/css/CSSPropertyInitialValues.cpp: Added.
* Source/WebCore/css/CSSPropertyInitialValues.h: Added.
    - Moved from CSSPropertyParser.h/cpp

* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StylePropertiesInlines.h:
    - Add #include for CSSPropertyInitialValues.h

* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
    - Removes all member data and functions, instead using static functions
      in CSSPropertyParser.cpp.

* Source/WebCore/css/parser/CSSPropertyParserCustom.h: Added.
    - Shorthand consumers moved here from CSSPropertyParser.cpp so they
      be called from generated dispatch function.

* Source/WebCore/css/parser/CSSPropertyParserResult.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserResult.h: Added.
    - Encapsulates the parser result vector and mutation functions.

* Source/WebCore/css/scripts/process-css-properties.py:
    - Support generating the shorthand parser dispatch. Utilize the new
      CSSPropertyParserCustom.h to also include all the parser helper
      headers, allowing us to avoid needing to update the script when
      it changes.

* Source/WebCore/style/StyleScopeRuleSets.cpp:
    - Add needed #include.

* Source/WebCore/css/scripts/test/TestCSSProperties.json:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.gperf:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyNames.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSPropertyParsing.h:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/CSSStyleProperties+PropertyNames.idl:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleBuilderGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleExtractorGenerated.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StyleInterpolationWrapperMap.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.cpp:
* Source/WebCore/css/scripts/test/TestCSSPropertiesResults/StylePropertyShorthandFunctions.h:
    - Update tests and results for new shorthand parser support properties and generation.

Canonical link: https://commits.webkit.org/295311@main
